velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -291,7 +291,7 @@ HiveWriterParameters::UpdateMode HiveDataSink::getUpdateMode() const {
           VELOX_UNSUPPORTED("Unsupported insert existing partitions behavior.");
       }
     } else {
-      VELOX_USER_FAIL("Unpartitioned Hive tables are immutable.");
+      return HiveWriterParameters::UpdateMode::kAppend;
```

Comment:
This behavior is controlled by "hive.immutable-partitions" configuration property. You need to check it before making it the append mode. I believe it should be called `HiveConfig::IsImmutablePartitions`. For the implementation logic, you can refer to Presto HiveWriterFactory
```
    private WriterParameters getWriterParameters(Optional<String> partitionName, OptionalInt bucketNumber)
    {
        // new table
        if (table == null) {
            // partitioned
            if (partitionName.isPresent()) {
                return getWriterParametersForNewPartitionedTable(partitionName.get());
            }
            // unpartitioned
            return getWriterParametersForNewUnpartitionedTable();
        }

        // existing unpartitioned table
        if (!partitionName.isPresent()) {
            return getWriterParametersForExistingUnpartitionedTable(bucketNumber);
        }

        // existing partitioned table
        return getWriterParametersForExistingPartitionedTable(partitionName.get(), bucketNumber);
    }

    private WriterParameters getWriterParametersForExistingUnpartitionedTable(OptionalInt bucketNumber)
    {
        // Note: temporary table is always empty at this step
        if (!table.getTableType().equals(TEMPORARY_TABLE)) {
            if (bucketNumber.isPresent()) {
                throw new PrestoException(HIVE_PARTITION_READ_ONLY, "Cannot insert into bucketed unpartitioned Hive table");
            }
            if (immutablePartitions) {
                throw new PrestoException(HIVE_PARTITION_READ_ONLY, "Unpartitioned Hive tables are immutable");
            }
        }
        return new WriterParameters(
                UpdateMode.APPEND,
                getHiveSchema(table),
                locationService.getTableWriteInfo(locationHandle),
                fromHiveStorageFormat(tableStorageFormat));
    }
```

Reply:
I have added the config to control this in Velox.